### PR TITLE
ruff rule UP007: Use X | Y for type annotations

### LINKED ISSUE #8327: ruff rule UP007: Use X | Y for type annotations for solr
Per https://github.com/internetarchive/openlibrary/pull/8236#discussion_r1333768598 we should be able to use those type annotations in solr now.

Will get the codebase to a more uniform style!

### Proposal & Constraints

Remove the solr exclusions here and add the proper annotations.
https://github.com/internetarchive/openlibrary/pull/8252/files#diff-50c86b7ed8ac2cf95bd48334961bf0530cdc77b5a56f852c5c61b89d735fd711

### Stakeholders

@cclauss would you like to do this?

### PR DESCRIPTION
**Pull request by cclauss**
Closes #8327

https://docs.astral.sh/ruff/rules/non-pep604-annotation

% `ruff --select=UP007 --fix .`
```
Found 99 errors (97 fixed, 2 remaining).
```
% `ruff rule UP007`
# non-pep604-annotation (UP007)

Derived from the **pyupgrade** linter.

Autofix is sometimes available.

## What it does
Check for type annotations that can be rewritten based on [PEP 604] syntax.

## Why is this bad?
[PEP 604] introduced a new syntax for union type annotations based on the
`|` operator. This syntax is more concise and readable than the previous
`typing.Union` and `typing.Optional` syntaxes.

This rule is enabled when targeting Python 3.10 or later (see:
[`target-version`]). By default, it's _also_ enabled for earlier Python
versions if `from __future__ import annotations` is present, as
`__future__` annotations are not evaluated at runtime. If your code relies
on runtime type annotations (either directly or via a library like
Pydantic), you can disable this behavior for Python versions prior to 3.10
by setting [`pyupgrade.keep-runtime-typing`] to `true`.

## Example
```python
from typing import Union

foo: Union[int, str] = 1
```

Use instead:
```python
foo: int | str = 1
```

## Options
- `target-version`
- `pyupgrade.keep-runtime-typing`

[PEP 604]: https://peps.python.org/pep-0604/

### CODE REVIEW

File: pyproject.toml [hunk 1]
```diff
@@ -175,10 +175,8 @@ max-statements = 70
 "openlibrary/plugins/upstream/borrow.py" = ["BLE001", "E722"]
 "openlibrary/plugins/upstream/models.py" = ["BLE001"]
 "openlibrary/plugins/upstream/utils.py" = ["BLE001"]
-"openlibrary/solr/*" = ["UP007"]
+"openlibrary/solr/solr_types.py" = ["UP007"]
```
Comment (RayBB) [inline pyproject.toml:178]: What stops us from migrating `openlibrary/solr/solr_types.py` as well?

**Review by RayBB — APPROVED** [review]
Spun up the PR in gitpod and the search and such worked as expected. Also all the changes look correct to me. Just one question about `openlibrary/solr/solr_types.py`.

File: pyproject.toml (hunk 1 above)
Comment (cdrini) [reply pyproject.toml:178]: Nothing ! That file is auto-generated though, so we'd need to change the file that generates it ( `types_generator.py` )

**Review by cdrini — APPROVED** [review]
Lgtm! I believe cython/solr pipelines support this syntax , but if they don't it'll only cause an issue during full reindex, and I can easily address that.

File: pyproject.toml (hunk 1 above)
Comment (cclauss) [reply pyproject.toml:178]: And the two `# noqa: UP007` in `scripts/solr_builder/tests/test_fn_to_cli.py`

**Pull request merged**